File: gateway/src/main/java/example/demo/filter/AutherizationGatewayFilterFactory.java
package example.demo.filter;

import auth.dto.AuthRequest;
import example.demo.client.AuthWebService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cloud.gateway.filter.GatewayFilter;
import org.springframework.cloud.gateway.filter.factory.AbstractGatewayFilterFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;


@Slf4j
@Component
@RequiredArgsConstructor
public class AutherizationGatewayFilterFactory extends AbstractGatewayFilterFactory {
    private final AuthWebService authWebService;

    @Override
    public GatewayFilter apply(Object config) {
        return (exchange, chain) -> {
            log.info("Auth filter request");
            return authRequest(exchange).flatMap(authRequest -> {
                        log.info("authorization redirect {}", authRequest.isAuth() ? "not required" : "required");
                        return authRequest.isAuth() ? chain.filter(exchange) : redirectToAutPage(exchange);
                    }
            );
        };
    }

    private Mono<AuthRequest> authRequest(ServerWebExchange exchange) {
        return authWebService.auth(exchange.getRequest().getQueryParams().get("user").get(0));

    }

    private Mono<Void> redirectToAutPage(ServerWebExchange exchange) {
        log.info("send redirect to authorization page");

        return Mono.fromRunnable(() -> {
            ServerHttpResponse response = exchange.getResponse();
            response.setStatusCode(HttpStatus.FOUND);
        });
    }

}

File: gateway/src/main/java/example/demo/filter/AutherizationGatewayFilterFactory.kt
package example.demo.filter

import auth.dto.AuthRequest
import auth.dto.AuthResponse
import auth.dto.Response
import example.demo.client.AuthorizationService
import example.demo.model.AuthContext
import mu.KotlinLogging
import org.springframework.cloud.gateway.filter.GatewayFilter
import org.springframework.cloud.gateway.filter.GatewayFilterChain
import org.springframework.cloud.gateway.filter.factory.AbstractGatewayFilterFactory
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebSession
import reactor.core.publisher.Mono
import java.util.*

@Component
class AuthorizationGatewayFilterFactory(val authorizationService: AuthorizationService) : AbstractGatewayFilterFactory<Any?>() {
    private val logger = KotlinLogging.logger {}

    override fun apply(config: Any?): GatewayFilter {
        return GatewayFilter { exchange: ServerWebExchange, chain: GatewayFilterChain ->
            logger.info("Auth filter request")
            authRequest(exchange).flatMap { authResponse : AuthResponse ->
                logger.info("authorization redirect {}", if (authResponse.auth) "not required" else "required")
                if (authResponse.auth) chain.filter(exchange) else redirectToAutPage(exchange)
            }
        }
    }

    private fun authRequest(exchange: ServerWebExchange): Mono<AuthResponse> {
        return exchange.session
                .flatMap {
                    val authContext = it.getAttribute<AuthContext>("auth")
                    return@flatMap if (authContext == null)
                        createInitRequest(it)
                    else {
                        authorizationService.getAuthCode(AuthRequest(authContext.userId));
                    }
                }
    }

    private fun redirectToAutPage(exchange: ServerWebExchange): Mono<Void> {
        logger.info("send redirect to authorization page")
        return Mono.fromRunnable {
            val response = exchange.response
            response.statusCode = HttpStatus.FOUND
        }
    }

    private fun createInitRequest(webSession: WebSession): Mono<AuthResponse> {
        webSession.attributes["auth"] = AuthContext(UUID.randomUUID().toString())
        return Mono.just(AuthResponse(false))

    }
}